packages/flutter/lib/src/semantics/binding.dart

```diff
@@ -28,12 +33,124 @@ mixin SemanticsBinding on BindingBase {
+  /// in them. These clients express their interest by holding a
+  /// [SemanticsHandle].
+  ///
+  /// Clients can close their [SemanticsHandle] by calling
+  /// [SemanticsHandle.dispose]. Once all outstanding [SemanticsHandle] objects
+  /// are closed, semantics information are no longer collected.
+  SemanticsHandle ensureSemantics() {
+    assert(_outstandingHandles >= 0);
+    _outstandingHandles++;
+    assert(_outstandingHandles > 0);
+    _semanticsEnabled.value = true;
+    return SemanticsHandle._(_didDisposeSemanticsHandle);
+  }
+
+  void _didDisposeSemanticsHandle() {
+    assert(_outstandingHandles > 0);
+    _outstandingHandles--;
+    assert(_outstandingHandles >= 0);
+    _semanticsEnabled.value = _outstandingHandles > 0;
+  }
+
+  // Handle for semantics request from the platform.
+  SemanticsHandle? _semanticsHandle;
+
+  void _handleSemanticsEnabledChanged() {
+    if (platformDispatcher.semanticsEnabled) {
+      _semanticsHandle ??= ensureSemantics();
+    } else {
+      _semanticsHandle?.dispose();
+      _semanticsHandle = null;
+    }
+  }
+
+  void _handleSemanticsAction(int id, ui.SemanticsAction action, ByteData? args) {
+    performSemanticsAction(SemanticsActionEvent(
+      nodeId: id,
+      type: action,
+      arguments: args != null ? const StandardMessageCodec().decodeMessage(args) : null,
+    ));
+  }
+
+  /// Called  whenever the platform requests an action to be performed on a
+  /// [SemanticsNode].
+  ///
+  /// This callback is invoked when a user interacts with the app via an
+  /// accessibility service (e.g. TalkBack and VoiceOver) and initiates an
+  /// action on the focused node.
+  ///
+  /// Bindings that mixin the [SemanticsBinding] must implement this method and
+  /// perform the given `action` on the [SemanticsNode] specified by
+  /// [SemanticsActionEvent.nodeId].
+  ///
+  /// See [dart:ui.PlatformDispatcher.onSemanticsAction].
+  @protected
+  void performSemanticsAction(SemanticsActionEvent action);
+
+  void _handleAccessibilityFeaturesChanged() {
+    _accessibilityFeatures = platformDispatcher.accessibilityFeatures;
+    handleAccessibilityFeaturesChanged();
+  }
+
+  /// The currently active set of [AccessibilityFeatures].
+  ///
+  /// This is set when the binding is first initialized and updated whenever a
+  /// flag is changed.
+  ///
+  /// To listen to changes to accessibility features, create a
+  /// [WidgetsBindingObserver] and listen to
+  /// [WidgetsBindingObserver.didChangeAccessibilityFeatures].
+  ui.AccessibilityFeatures get accessibilityFeatures => _accessibilityFeatures;
+  late ui.AccessibilityFeatures _accessibilityFeatures;
+
   /// Called when the platform accessibility features change.
   ///
   /// See [dart:ui.PlatformDispatcher.onAccessibilityFeaturesChanged].
   @protected
   void handleAccessibilityFeaturesChanged() {
-    _accessibilityFeatures = platformDispatcher.accessibilityFeatures;
+    // Nothing to do by default.
```

Comment:
Why not just mark this mustCallSuper?

Reply:
My thinking was that this setup guarantees that `accessibilityFeatures` is called before handleAccessibilityFeaturesChanged is invoked on subclasses. With `mustCallSuper` you could accidentally call super at the end of your method and operate on outdated data.

Reply:
But we don't do it this way for other handlers on the bindings, so I will change this to your suggestion as well.